controllers/vspherecluster_reconciler.go

```diff
@@ -358,15 +360,29 @@ func (r clusterReconciler) reconcileDeploymentZones(ctx *context.ClusterContext)
 	failureDomains := clusterv1.FailureDomains{}
 	for _, zone := range deploymentZoneList.Items {
 		if zone.Spec.Server == ctx.VSphereCluster.Spec.Server {
+			// This should never fail, validating webhook should catch this first
+			if ctx.VSphereCluster.Spec.ZoneSelector != nil {
+				selector, err := metav1.LabelSelectorAsSelector(ctx.VSphereCluster.Spec.ZoneSelector)
+				if err != nil {
+					return false, errors.Wrapf(err, "zone label selector is misconfigured")
+				}
+
+				// An empty selector allows the zone to be selected
+				if !selector.Empty() && !selector.Matches(labels.Set(zone.GetLabels())) {
+					r.Logger.V(5).Info("skipping the deployment zone due to label mismatch", "name", zone.Name)
+					continue
+				}
+			}
+
 			if zone.Status.Ready == nil {
```

Comment:
Probably not related to this PR, but...
Based on this logic, if any of the "available deployment zone" is not ready, it will fail the whole deployment. Now, if I have 6 deployment zones, and a request of 3 replicas, but of this 6 zones 1 is off, wouldn't this stop the whole deployment anyway?

Reply:
The idea is to wait until all the zones have reported their status. Irrespective of the ready status (`true` or `false`), as long as it is reported, we let the reconciliation pass through. As for including the zones in the failure domain list, only the zone with Ready reported as `true` are included in the list.